File: packages/service-template/templates/python/src/adapters/in/health_listener.py
import json
import os
import signal
import socket
import sys
import threading
from dataclasses import asdict

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")))
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "ports", "in")))

from health_check_usecase_port_in import HealthCheckUseCasePortIn


class HealthListener:
    """Liveness probe only — does not parse the request or route by path."""

    def __init__(self, usecase: HealthCheckUseCasePortIn, stop_event: threading.Event) -> None:
        self._usecase = usecase
        self._stop_event = stop_event
        signal.signal(signal.SIGTERM, self._handle_signal)
        signal.signal(signal.SIGINT, self._handle_signal)
        host = os.environ.get("HEALTH_HOST", "0.0.0.0")
        port = int(os.environ.get("HEALTH_PORT", "8080"))
        self._server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._server.bind((host, port))
        self._server.listen(5)
        self._server.settimeout(1.0)

    def _handle_signal(self, signum: int, frame: object) -> None:
        self._stop_event.set()

    def start(self) -> None:
        try:
            while not self._stop_event.is_set():
                try:
                    conn, _ = self._server.accept()
                except OSError:
                    continue
                with conn:
                    conn.settimeout(2.0)
                    try:
                        conn.recv(1024)
                        body = json.dumps(asdict(self._usecase.check())).encode()
                        response = (
                            b"HTTP/1.1 200 OK\r\n"
                            b"Content-Type: application/json\r\n"
                            b"Content-Length: " + str(len(body)).encode() + b"\r\n"
                            b"Connection: close\r\n\r\n" + body
                        )
                        conn.sendall(response)
                    except OSError:
                        pass
        finally:
            self._server.close()

File: packages/service-template/templates/rust/src/adapters/in/health_listener.rs
use std::io::{Read, Write};
use std::net::TcpListener;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use crate::ports::port_in::health_check_usecase_port_in::HealthCheckUseCasePortIn;

pub struct HealthListener {
    listener: TcpListener,
    usecase: Box<dyn HealthCheckUseCasePortIn>,
    stop_flag: Arc<AtomicBool>,
}

impl HealthListener {
    pub fn new(usecase: Box<dyn HealthCheckUseCasePortIn>, stop_flag: Arc<AtomicBool>) -> Self {
        signal_hook::flag::register(signal_hook::consts::SIGTERM, stop_flag.clone())
            .expect("failed to register SIGTERM handler");
        signal_hook::flag::register(signal_hook::consts::SIGINT, stop_flag.clone())
            .expect("failed to register SIGINT handler");
        let host = std::env::var("HEALTH_HOST").unwrap_or_else(|_| "0.0.0.0".to_string());
        let port: u16 = std::env::var("HEALTH_PORT")
            .unwrap_or_else(|_| "8080".to_string())
            .parse()
            .expect("HEALTH_PORT must be a valid port number");
        let listener = TcpListener::bind((host.as_str(), port)).expect("failed to bind health port");
        listener
            .set_nonblocking(true)
            .expect("failed to set health listener nonblocking");
        Self { listener, usecase, stop_flag }
    }

    /// Liveness probe only — does not parse the request or route by path.
    pub fn start(&self) {
        while !self.stop_flag.load(Ordering::SeqCst) {
            match self.listener.accept() {
                Ok((mut stream, _)) => {
                    // The accepted connection inherits the listener's non-blocking
                    // flag — must be reset before writing, or write_all can return
                    // WouldBlock mid-response and truncate it.
                    if let Err(e) = stream.set_nonblocking(false) {
                        eprintln!("__SERVICE_NAME__: health listener set_nonblocking(false) failed: {e}");
                        continue;
                    }
                    // Drain (a bounded slice of) the client's request before responding.
                    // Dropping the stream while its request bytes are still unread makes
                    // the OS send RST instead of a clean FIN on close, and that RST can
                    // discard the response we already wrote before the client's kernel
                    // reads it out of its own receive buffer — truncating the body.
                    let _ = stream.set_read_timeout(Some(std::time::Duration::from_millis(500)));
                    let mut drain_buf = [0u8; 1024];
                    let _ = stream.read(&mut drain_buf);
                    let body = serde_json::to_string(&self.usecase.check())
                        .expect("failed to serialize health status");
                    let response = format!(
                        "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
                        body.len()
                    );
                    // A slow or dead client (e.g. one that connects but never reads)
                    // could otherwise block write_all forever, hanging this thread
                    // and every future health check with it.
                    let _ = stream.set_write_timeout(Some(std::time::Duration::from_secs(2)));
                    if let Err(e) = stream.write_all(response.as_bytes()) {
                        eprintln!("__SERVICE_NAME__: health listener header write failed: {e}");
                        continue;
                    }
                    if let Err(e) = stream.write_all(body.as_bytes()) {
                        eprintln!("__SERVICE_NAME__: health listener body write failed: {e}");
                    }
                }
                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                    std::thread::sleep(std::time::Duration::from_millis(200));
                }
                Err(e) => eprintln!("__SERVICE_NAME__: health listener accept failed: {e}"),
            }
        }
    }
}

